### PR TITLE
Added low memory configuration option to CMake and configure.

### PR DESCRIPTION
**Pull request by nmoinvaz**
This is based on discussion from #868.

### CODE REVIEW

**Comment by gdevenyi** [comment]
Should this option perhaps provide a definition of what low-memory means in the zlib context? I doubt it means a 1GB VM for example?

File: configure [hunk 1]
```diff
@@ -155,6 +156,7 @@ case "$1" in
       echo '    [--without-neon]            Compiles without ARM Neon SIMD instruction set' | tee -a configure.log
       echo '    [--with-dfltcc-deflate]     Use DEFLATE CONVERSION CALL instruction for compression on IBM Z' | tee -a configure.log
       echo '    [--with-dfltcc-inflate]     Use DEFLATE CONVERSION CALL instruction for decompression on IBM Z' | tee -a configure.log
+      echo '    [--with-low-mem]            Build for low memory environments (reduced performance)' | tee -a configure.log
```
Comment (Dead2) [inline configure]: Would it be clearer if the text was something like this?
"Reduced memory usage for special cases (reduces performance)"

Comment (Dead2) [reply configure]: Likewise for the other messages of course.

Comment (Dead2) [reply configure]: @nmoinvaz ^

Comment (nmoinvaz) [reply configure]: Sorry. I'm not really a fan of this option but I can continue to add it. 

Should I rename the option to `--with-reduced-mem` too?

Comment (Dead2) [reply configure]: Oh, I didn't know that.

I agree that this option should generally not be used except in certain cases (such as a very very busy nginx proxy for example). It is an useful option, just for a very narrow few niches.

So I think that it should be tagged as an advanced option in cmake.
And perhaps it should not show in the summary unless enabled (Also goes for a few of the other options we'd rather not people use unless they really know what they are doing).

Comment (Dead2) [reply configure]: And yes, it would be smart to rename the option as well 😄

**Comment by nmoinvaz** [comment]
Rebased.

File: configure (hunk 1 above)
Comment (nmoinvaz) [reply configure]: Done.

**Comment by nmoinvaz** [comment]
@Dead2 this has problem in CI. `WITH_REDUCED_MEM` is enabled by default.

File: CMakeLists.txt [hunk 2]
```diff
@@ -76,6 +76,7 @@ option(ZLIB_ENABLE_TESTS "Build test binaries" ON)
 option(ZLIB_DUAL_LINK "Dual link tests against system zlib" OFF)
 option(WITH_FUZZERS "Build test/fuzz" OFF)
 option(WITH_OPTIM "Build with optimisation" ON)
+option(WITH_REDUCED_MEM "Reduced memory usage for special cases (reduces performance)" ON)
```
Comment (nmoinvaz) [inline CMakeLists.txt:79]: @Dead2 This should be set to OFF.

**Comment by Dead2** [comment]
@nmoinvaz Good thing it only went to develop so far then 😄, are you making a PR to fix it?

**Comment by nmoinvaz** [comment]
Yes I can

**Pull request merged**